rockcraft/services/package.py

```diff
@@ -133,14 +135,47 @@ def _pack(
         emit.progress(f"Setting the default OCI user to be {project.run_user}")
         new_image.set_default_user(userid, project.run_user)
 
-    emit.progress("Adding Pebble entrypoint")
-
-    new_image.set_entrypoint(
-        project.entrypoint_service, project.build_base or project.base
-    )
-    if project.services and project.entrypoint_service in project.services:
-        new_image.set_cmd(project.services[project.entrypoint_service].command)
-
+    if project.entrypoint_command:
+        emit.progress("Setting OCI entrypoint")
+
+        entrypoint: list[str] = []
+        cmd: list[str] = []
+
+        # Check arguments
+        in_brackets = False
+        for arg in shlex.split(project.entrypoint_command):
+            if arg == "[":
+                in_brackets = True
+                continue
+            if arg == "]":
+                in_brackets = False
+                continue
+            (cmd if in_brackets else entrypoint).append(arg)
+
+    else:
+        emit.progress("Adding Pebble entrypoint")
+
+        entrypoint = Pebble.get_entrypoint(project.build_base or project.base)
+        cmd = []
+
+        if project.entrypoint_service:
+            entrypoint.extend(["--args", project.entrypoint_service])
+
+        if project.services and project.entrypoint_service in project.services:
+            command = project.services[project.entrypoint_service].command
+            command_sh_args = shlex.split(command or "")
```

Comment:
I think we have this logic of splitting a string like `outside [ inside ]` into `([outside], [inside])` in at least three different places by now. Can we isolate that in a separate function please?